src/gc/collector.cpp

```diff
@@ -333,50 +372,62 @@ void markPhase() {
-                if (global_heap.small_arena.contains(p)) {
-                    SmallArena::Block* b = SmallArena::Block::forPointer(p);
-                    assert(b->size >= bytes + sizeof(GCAllocation));
-                }
-            }
-            visitor.visitPotentialRange((void**)p, (void**)((char*)p + bytes));
-        } else if (kind_id == GCKind::PRECISE) {
-            uint32_t bytes = al->kind_data;
-            if (DEBUG >= 2) {
-                if (global_heap.small_arena.contains(p)) {
-                    SmallArena::Block* b = SmallArena::Block::forPointer(p);
-                    assert(b->size >= bytes + sizeof(GCAllocation));
-                }
-            }
-            visitor.visitRange((void**)p, (void**)((char*)p + bytes));
-        } else if (kind_id == GCKind::PYTHON) {
-            Box* b = reinterpret_cast<Box*>(p);
-            BoxedClass* cls = b->cls;
-
-            if (cls) {
-                // The cls can be NULL since we use 'new' to construct them.
-                // An arbitrary amount of stuff can happen between the 'new' and
-                // the call to the constructor (ie the args get evaluated), which
-                // can trigger a collection.
-                ASSERT(cls->gc_visit, "%s", getTypeName(b));
-                cls->gc_visit(&visitor, b);
-            }
-        } else if (kind_id == GCKind::HIDDEN_CLASS) {
-            HiddenClass* hcls = reinterpret_cast<HiddenClass*>(p);
-            hcls->gc_visit(&visitor);
-        } else {
-            RELEASE_ASSERT(0, "Unhandled kind: %d", (int)kind_id);
+    long us = _t.end();
+    sc_us.log(us);
+}
+
+static void markPhase() {
+    static StatCounter sc_us("us_gc_mark_phase");
+    Timer _t("markPhase", /*min_usec=*/10000);
+
+#ifndef NVALGRIND
+    // Have valgrind close its eyes while we do the conservative stack and data scanning,
+    // since we'll be looking at potentially-uninitialized values:
+    VALGRIND_DISABLE_ERROR_REPORTING;
+#endif
+
+#if TRACE_GC_MARKING
+#if 1 // separate log file per collection
+    char tracefn_buf[80];
+    snprintf(tracefn_buf, sizeof(tracefn_buf), "gc_trace_%03d.txt", ncollections);
+    trace_fp = fopen(tracefn_buf, "w");
+#else // overwrite previous log file with each collection
+    trace_fp = fopen("gc_trace.txt", "w");
+#endif
+#endif
+    GC_TRACE_LOG("Starting collection %d\n", ncollections);
+
+    GC_TRACE_LOG("Looking at roots\n");
+    TraceStack stack(roots);
+    GCVisitor visitor(&stack);
+
+    markRoots(visitor);
+
+    graphTraversalMarking(stack, visitor);
+
+    // Some classes might be unreachable. Unfortunately, we have to keep them around for
+    // one more collection, because during the sweep phase, instances of unreachable
+    // classes might still end up looking at the class. So we visit those unreachable
+    // classes and remove it from the list of class objects so that it can be freed
+    // in the next collection.
+    std::vector<BoxedClass*> classes_to_remove;
```

Comment:
it might be more efficient to create a new `class_objects` vector here (maybe reserve the same size as the old `class_objects` vector to save allocations), and only push_back the classes onto the new one if they're marked.